### PR TITLE
Add support for milliseconds formatted as %f in strptime(fmt).

### PR DESCRIPTION
**Pull request by reidwagner**
Addresses #1409.

The milliseconds should be added at the end of the fmt string. They're chopped off of the original fmt string which is passed to strptime(3). As before, if the format does not match, e.g. milliseconds in date but not fmt string, a jv error is returned. 

I also moved some of the cleanup in case of an error to a goto label to reduce redundancy. 

Please let me know if I can change or improve anything!

### CODE REVIEW

File: src/builtin.c [hunk 1]
```diff
@@ -1219,24 +1219,49 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
-  if (end == NULL || (*end != '\0' && !isspace(*end))) {
-    jv e = jv_invalid_with_msg(jv_string_fmt("date \"%s\" does not match format \"%s\"", input, fmt));
-    jv_free(a);
-    jv_free(b);
-    return e;
+  char *inputdup = strdup(input);
+  char *fmtdup = strdup(fmt);
+
+  int ms = -1;
+  char *pf = strstr(fmtdup, "%f");
```
Comment (nicowilliams) [inline src/builtin.c]: Watch out! `%` can be escaped with a preceding (unescaped) `%`.  You have to step over the format string character (well, byte) by character (byte).

**Comment by nicowilliams** [comment]
I wonder if we may not have to implement `strptime()` from scratch, rather than depend on the OS `strptime()`.  That may be advisable as to all system date functions, since they tend to vary somewhat by OS.  That would be painful though, especially as it means having to deal with timezones, which we should absolutely not want to do!  So I guess we'll hack it somewhat like your PR.

File: src/builtin.c [hunk 2]
```diff
@@ -1219,24 +1219,49 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+  if (pf) {
+    char *msfmtloc = pf - 1;
```
Comment (nicowilliams) [inline src/builtin.c]: Buffer underrun if the format string starts with `"%f"`.

File: src/builtin.c [hunk 3]
```diff
@@ -1219,24 +1219,49 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+    char delim = *msfmtloc;
+    *msfmtloc = '\0';
+    char *msloc = strrchr(inputdup, delim);
```
Comment (nicowilliams) [inline src/builtin.c]: The assumption here is that `delim` will not be used for anything else after the `%f`...  I'm not very comfortable with this.

**Comment by nicowilliams** [comment]
Note that `strptime()` returns a pointer to the first character not consumed by `strptime()`.  So I think what we want to do is something like this:

 - implement a function that supports all the format string features that `strptime()` supports and which half-parses the input string, only skipping each field as it goes
 - if and when it reaches `%f` this function should
    - parse the field
    - construct new format and input strings with the `%f` and milliseconds field removed, respectively
 - call the real `strptime()` with the possibly-new format and input strings to do the real parsing

That's a lot of work.  In the mean time users can use regular expressions to parse and fixup timestamps.

Lastly, `%f`... is that used by any implementation anywhere?  We don't want to conflict with C library implementations nor POSIX...

**Comment by nicowilliams** [comment]
We're in luck as to `%f`: it's already in use in various places to mean "milliseconds" (e.g., Python 2.6 and up).

**Comment by nicowilliams** [comment]
(Hit wrong button; didn't mean to close this!)

**Comment by nicowilliams** [comment]
Bah, Ruby uses `%L` and `%N`.  R uses `%OSn` (which conflicts with glibc!).

What a mess.

**Comment by rwagnerr** [comment]
Thanks for the fast response. I'm going to try working out the plan you proposed and address the code comments.

**Comment by c0b** [comment]
have progress since May?

**Comment by reidwagner** [comment]
I haven't gone any further.

**Comment by nicowilliams** [comment]
I'd definitely like to complete this.  I can't promise that I will soon though :(

**Comment by spier** [comment]
I just also ran into this issue with parsing milliseconds in jq.
Is there anything that I could do to help with this PR?

**Comment by reidwagner** [comment]
I'm taking a fresh look at this using the plan put forth by @nicowilliams.

**Comment by reidwagner** [comment]
I was hoping to find a clever way to simply parse out `%f` from the format string without paying attention to the values of each field. I.e. determining the delimiters and matching fields until `%f`. And you can get pretty close making some assumptions. But there's no two ways around something like `"Februaryr%Br%Y"`. So yeah, it would have to support every feature as you said.

Because I probably don't have the bandwidth for that I'm going to close this to encourage anyone else who wants to try.

**Pull request closed without merging**